### PR TITLE
Add 'include_leased' flag to ListBalances

### LINKED ISSUE #1106: [feature]: Enhance `asset_sum` report to clarify conservation
## Background
Strict, asset-amount conservation checks exists in the tapd VM to prevent undue inflation of asset quantities. However, there may be instances where the daemon reports asset amounts that appear to be double counted, even though they are being properly conserved.

### Asset Balance Clarification

To avoid confusion, it is essential to clearly delineate between payment-channel balances and L1-only asset balances. This will ensure that users have a comprehensive understanding of their asset holdings.

#### Task 1: Ensure Mutual Exclusivity
- [ ] Verify that the sum of L1-only asset balances and payment-channel balances are mutually exclusive, i.e., there is
 no overlap or double counting between the two types of balances.
  - [ ] Ensure that asset balance sum reports, sum(non-channel asset balances + local_balance)

### Comprehensive Asset Sum Calculations

To ensure accuracy in asset sum calculations, we need to investigate the comprehensiveness of these calculations.

#### Task 2: Investigate `asset sum` Calculation
- [x] https://github.com/lightninglabs/taproot-assets/issues/1106#issuecomment-2391574186 ~~Check if the `script_key_has_script_path": true` flag is being incorrectly included in the asset sum. (Intuition is no)~~
  - [x] In situ: Assets have been incorporated in the db (due to some other reason).  Confirm that these assets are not being counted erroneously in asset sum.
- [x] Verify that leased assets are not showing up in the asset sum, as they may be subject to different conservation rules. [Completed in https://github.com/lightninglabs/taproot-assets/pull/1119] 

## Background
In the scenario where two nodes have a channel
Node 1 force closes
Node 2 (the responder) -- even if it didn't have the channel's entire balance, the node insert the force close transaction, into it's own database
These inserted transaction are added to the database fields that are importing the outputs as well
While these outputs cannot be spent due to the presence of a script path.
But litd/tapd displays these as part of the node's balance today

## Solution
`make another delineated name space, to track the amounts` in order for the `wallet balance commands` to `by default not show these balance, by default`

## Considerations
These values are used during tests to assert certain outcomes have been met
`add a flag to enable reporting` in order for `assertion tests` to `have the necessary data available to run tests`

### PR DESCRIPTION
**Pull request by gijswijs**
This PR adds `include_leased` flag to ListBalances rpc endpoint.
With this flag you show balances that include the leased assets, which are otherwise ignored. In doing so the behavior is now synced with the `ListAssets` endpoint.

Fixes #1106 task 2 (2nd checkbox, and probably also 1st checkbox)

Fixes https://github.com/lightninglabs/taproot-assets/issues/1106

### CODE REVIEW

File: cmd/tapcli/assets.go [hunk 1]
```diff
@@ -693,6 +698,8 @@ func listAssetBalances(ctx *cli.Context) error {
 		}
 	}
 
+	req.IncludeLeased = ctx.Bool(assetIncludeLeasedName)
```
Comment (guggero) [inline cmd/tapcli/assets.go:701]: nit: could just add this to where the struct is initialized:
```
	req := &taprpc.ListBalancesRequest{
		IncludeLeased: ctx.Bool(assetIncludeLeasedName),
	}
```

File: rpcserver.go [hunk 2]
```diff
@@ -938,7 +938,7 @@ func (r *rpcServer) checkBalanceOverflow(ctx context.Context,
 	case groupPubKey != nil:
 		// Retrieve the current balance of the group.
 		balances, err := r.cfg.AssetStore.QueryAssetBalancesByGroup(
-			ctx, groupPubKey,
+			ctx, groupPubKey, false,
```
Comment (guggero) [inline rpcserver.go:941]: Same here, I think we'll want `true`.

File: tapdb/assets_store.go [hunk 3]
```diff
@@ -1014,20 +1042,37 @@ func (a *AssetStore) QueryBalancesByAsset(ctx context.Context,
...
+	// filtering our database query understands.
+	assetBalancesFilter := QueryAssetBalancesByGroupFilters{
+		Now: sql.NullTime{
+			Time:  a.clock.Now().UTC(),
+			Valid: true,
+		},
+	}
+
+	// By default, we only show assets that are not leased.
+	if !includeLeased {
```
Comment (guggero) [inline tapdb/assets_store.go:1059]: Would be nice to also add some unit tests for the flag, especially around the three different values for the `leased` flag on the DB level (`NULL` = no filter, `true`  = only leased, `false` = only non-leased).

File: tapdb/assets_store.go [hunk 4]
```diff
@@ -1014,20 +1042,37 @@ func (a *AssetStore) QueryBalancesByAsset(ctx context.Context,
...
+		assetBalancesFilter.Leased = sqlBool(false)
+	}
+
+	// Only show specific group if a groupKey has been passed
```
Comment (guggero) [inline tapdb/assets_store.go:1063]: nit: missing full stop.

File: itest/assertions.go [hunk 5]
```diff
@@ -1886,7 +1886,7 @@ func AssertGenesisOutput(t *testing.T, output *taprpc.ManagedUtxo,
 }
 
 func AssertAssetBalances(t *testing.T, client taprpc.TaprootAssetsClient,
-	simpleAssets, issuableAssets []*taprpc.Asset) {
+	simpleAssets, issuableAssets []*taprpc.Asset, includeLeased bool) {
```
Comment (guggero) [inline itest/assertions.go:1889]: Since we should now have the same flags for the `ListAssets` and `ListBalances` calls, I wonder if we should also add a check here that makes sure the sum of all assets returned by `ListAssets` matches the sum returned in `ListBalances`?

File: rpcserver.go [hunk 6]
```diff
@@ -922,7 +922,7 @@ func (r *rpcServer) checkBalanceOverflow(ctx context.Context,
 	case assetID != nil:
 		// Retrieve the current asset balance.
 		balances, err := r.cfg.AssetStore.QueryBalancesByAsset(
-			ctx, assetID,
+			ctx, assetID, false,
```
[the remark this answers is not in the document]
Comment (GeorgeTsagk) [reply rpcserver.go:925]: shouldn't we preserve the previous behavior? unless it was wrong before, for not including leased assets in the check

[the remark this answers is not in the document]
Comment (guggero) [reply rpcserver.go:925]: Yes, we want the previous behavior, which means include leased assets. Which would be `true` at this level (where it means `includeLeased`).

File: itest/assertions.go (hunk 5 above)
Comment (gijswijs) [reply itest/assertions.go:1889]: Your wish is my command.

File: tapdb/sqlc/queries/assets.sql [hunk 7]
```diff
@@ -321,7 +321,18 @@ JOIN genesis_info_view
 -- around that needs to be used with this query until a sqlc bug is fixed.
 LEFT JOIN key_group_info_view
     ON assets.genesis_id = key_group_info_view.gen_asset_id
-WHERE spent = FALSE
+LEFT JOIN managed_utxos utxos
```
[the remark this answers is not in the document]
Comment (gijswijs) [reply tapdb/sqlc/queries/assets.sql:324]: Correct, this was done on purpose. The initial approach was having a normal join with the `leased CASE` in the join clause, but this failed the `TestCommitBatchChainActions` unit test here: 
https://github.com/lightninglabs/taproot-assets/blob/c7fe6ef4afc9848cde8e19ee8e11f795fcaa0003/tapdb/asset_minting_test.go#L1044

The reason it failed is because that unit test expects to see the group genesis asset. But group genesis assets have `anchor_utxo_id` set to `NULL`, so they don't show up in the join unless it is a left join. Having the left join meant that I had to change the semantics of the query ever so slightly.

I now changed the join to have the semantics be consistent with `QueryAssets`while still making sure that the group genesis assets show up.

File: tapdb/assets_store.go (hunk 3 above)
Comment (gijswijs) [reply tapdb/assets_store.go:1059]: Added a unit test. Although the state 'only leased' can't be hit, because we never run the query with `leased = true`. 
```
if !includeLeased {
	assetBalancesFilter.Leased = sqlBool(false)
}
```
All other states are extensively tested with grouped and leased assets mixed.

File: tapdb/assets_store_test.go [hunk 8]
```diff
@@ -2205,3 +2205,147 @@ func TestTransferOutputProofDeliveryStatus(t *testing.T) {
...
 	)
 }
+
+func TestQueryAssetBalances(t *testing.T) {
+	t.Parallel()
+
+	_, assetsStore, _ := newAssetStore(t)
+	ctx := context.Background()
+
+	// First, we'll generate 3 assets, two of them sharing the same anchor
```
Comment (guggero) [inline tapdb/assets_store_test.go:2215]: nit: 4 assets?

File: tapdb/assets_store_test.go [hunk 9]
```diff
@@ -2205,3 +2205,147 @@ func TestTransferOutputProofDeliveryStatus(t *testing.T) {
...
+	for _, balance := range balancesByGroup {
+		balanceByGroupSum += balance.Balance
+	}
+	require.Equal(t, totalGroupedBalances, balanceByGroupSum)
+
+	// Now we lease the first asset for 1 hour. This will cause the second
+	// one also to be leased, since it's on the same anchor transaction. The
+	// second asset is in its own group, so when leased, the entire group is
+	// leased.
+	leaseOwner := fn.ToArray[[32]byte](test.RandBytes(32))
```
Comment (guggero) [inline tapdb/assets_store_test.go:2286]: :+1:

File: tapdb/sqlc/queries/assets.sql [hunk 10]
```diff
@@ -321,7 +321,18 @@ JOIN genesis_info_view
 -- around that needs to be used with this query until a sqlc bug is fixed.
 LEFT JOIN key_group_info_view
     ON assets.genesis_id = key_group_info_view.gen_asset_id
-WHERE spent = FALSE
+JOIN managed_utxos utxos
+    ON (assets.anchor_utxo_id = utxos.utxo_id OR assets.anchor_utxo_id IS NULL) AND
```
[the remark this answers is not in the document]
Comment (gijswijs) [reply tapdb/sqlc/queries/assets.sql:325]: The issue is that when anchoring pending assets with `AnchorPendingAssets` we select the assets on `prev_out`, but the genesis asset has a different `prev_out` than the other assets in the batch. So all assets _except_ the genesis asset is anchored.

[the remark this answers is not in the document]
Comment (gijswijs) [reply tapdb/sqlc/queries/assets.sql]: So `addRandGroupToBatch` creates a random asset genesis and stores it with `storeGroupGenesis`. But that random asset doesn't have a managed utxo in the database. It's inserted with `upsertAssetsWithGenesis` and although that func does allow for a `utxo_id` being passed, in this case `nil` is passed.

I have changed it, so that it _also_ creates a managed utxo in the database, and then passes the `utxo_id` it gets back from that operation to `upsertAssetsWithGenesis`. It's quite involved.

Because that random asset is now a proper asset that returns in `FetchAllAssets` it does change some of the assumptions in the original failing test `TestCommitBatchChainActions`.

What this all means is that we can have proper JOINS without the trickery from before.

File: tapdb/asset_minting.go [hunk 11]
```diff
@@ -1566,6 +1566,7 @@ func (a *AssetMintingStore) CommitSignedGenesisTx(ctx context.Context,
 			PrevOut:      genesisOutpoint,
 			AnchorUtxoID: sqlInt64(utxoID),
 		})
+
```
Comment (guggero) [inline tapdb/asset_minting.go:1569]: nit: unnecessary change?

File: tapdb/asset_minting_test.go [hunk 12]
```diff
@@ -149,10 +149,13 @@ func storeGroupGenesis(t *testing.T, ctx context.Context, initGen asset.Genesis,
 	upsertAsset := func(q PendingAssetStore) error {
 		_, err = maybeUpsertAssetMeta(ctx, q, &assetGen, nil)
 		require.NoError(t, err)
+		// Insert a random managed UTXO.
```
Comment (guggero) [inline tapdb/asset_minting_test.go:152]: nit: new line before comment.

File: tapdb/asset_minting_test.go [hunk 13]
```diff
@@ -168,6 +171,85 @@ func storeGroupGenesis(t *testing.T, ctx context.Context, initGen asset.Genesis,
 	}
 }
 
+// AddRandomManagedUTXO is a helper function that will create a random managed
```
Comment (guggero) [inline tapdb/asset_minting_test.go:174]: nit: Capitalization of Godoc function name doesn't match actual function name.

File: tapdb/asset_minting_test.go [hunk 14]
```diff
@@ -168,6 +171,85 @@ func storeGroupGenesis(t *testing.T, ctx context.Context, initGen asset.Genesis,
...
+	// Insert an internal key.
+	_, err = db.UpsertInternalKey(ctx, InternalKey{
+		RawKey:    randPubKey.SerializeCompressed(),
+		KeyFamily: 1,
+		KeyIndex:  2,
+	})
+	require.NoError(t, err)
+
+	// Insert the managed UTXO.
+	managedUTXO := RawManagedUTXO{
```
Comment (guggero) [inline tapdb/asset_minting_test.go:234]: Nice, this looks good!

File: tapdb/asset_minting_test.go [hunk 15]
```diff
@@ -790,6 +872,7 @@ type randAssetCtx struct {
 	tapSiblingBytes []byte
 	tapSiblingHash  chainhash.Hash
 	mintingBatch    *tapgarden.MintingBatch
+	randomGenesis   *asset.Genesis
```
Comment (guggero) [inline tapdb/asset_minting_test.go:875]: nit: since this is the group's genesis, should we call this `groupGenesis` instead?

File: tapdb/asset_minting_test.go [hunk 16]
```diff
@@ -967,13 +1051,14 @@ func TestCommitBatchChainActions(t *testing.T) {
 	)
 	require.Equal(t, txIndex, extractSqlInt32[uint32](dbGenTx.TxIndex))
 
-	// If we query for the set of all active assets, then we should get
-	// back the same number of seedlings.
+	// If we query for the set of all active assets, then we should get back
```
Comment (guggero) [inline tapdb/asset_minting_test.go:1054]: :+1:

File: tapdb/asset_minting_test.go [hunk 17]
```diff
@@ -995,9 +1080,15 @@ func TestCommitBatchChainActions(t *testing.T) {
 	// All the assets should also have a matching asset version as the
 	// seedlings we created.
 	mintingBatch := randAssetCtx.mintingBatch
+	randomGenesisTag := randAssetCtx.randomGenesis.Tag
 	require.True(t, fn.All(assets, func(dbAsset *asset.ChainAsset) bool {
 		seedling, ok := mintingBatch.Seedlings[dbAsset.Genesis.Tag]
 		if !ok {
+			// The only asset that doesn't have a seedling is the
```
Comment (guggero) [inline tapdb/asset_minting_test.go:1087]: :+1:

**Review by guggero — APPROVED** [review]
LGTM, pending a rebase/squash and some more nits :tada:

File: tapdb/asset_minting_test.go (hunk 13 above)
Comment (gijswijs) [reply tapdb/asset_minting_test.go:174]: Good catch. I had it first defined someplace else.

File: tapdb/asset_minting_test.go (hunk 14 above)
Comment (gijswijs) [reply tapdb/asset_minting_test.go:234]: :+1:

File: itest/assertions.go [hunk 18]
```diff
@@ -1913,7 +1913,9 @@ func AssertAssetBalances(t *testing.T, client taprpc.TaprootAssetsClient,
 
 	require.Equal(t, len(allAssets), len(assetIDBalances.AssetBalances))
 
+	var totalBalance uint64
```
Comment (GeorgeTsagk) [inline itest/assertions.go:1916]: nit: shouldn't this commit precede the previous `itest: test balances with an without leased flag`

**Review by GeorgeTsagk — APPROVED** [review]
Lgtm!
very nice commit structure and thorough unit test 💯
